Sem 146 no wifi worries mate imx8

robinmdh opened the pull request:
## Description
This is essentially a copy of #176 but form imx8
The edits during cherry-picking were rather small to be able to do this.
This is somewhat related to https://github.com/Ultimaker/opinicus/pull/2300 which also writes the eeprom but can deal with the crc checksum we use for imx 8 machines. Since we don't want to impact the workflow in the factory and so they can do both the locking and the article number/bom assignment at the same time saving a reboot.

The description stolen from that PR follows below:

_This PR is part of Ultimaker/opinicus#2003, and implements changes to the kernel boot script._

The current [printer provisioning process](https://ultimaker.atlassian.net/l/cp/oRBu1zNW) is documented quite extensively on our wiki. It basically boils down the following:

Early on in the boot process, the kernel init script (`initramfs/init.sh` in the repo) performs a bunch of tasks. One of these checks if a USB drive has been inserted. If that USB drive contains a single ext4 partition, it can be mounted. If the script then finds an `article_number` file in the drive  root, it programs the BOM number value in the mainboard EEPROM with the contents of the file.

We extend this by allowing an additional file to be present in the USB drive's root, `country_code_lock`. The contents of this file will be written to addres `0x0118` in the mainboard EEPROM, from where it will be picked up by the opinicus firmware in Ultimaker/opinicus#2003.

To provision a printer so that it's locked to Australia, you'd take a USB stick with a file `country_code_lock` that contains a single line of text:
```
0x41 0x55
```
To wipe the Australia lock from a printer, you'd put a file `country_code_lock` with a single line:
```
0xFF 0xFF
```
To provision a printer in a way that simply leaves the country lock as-is, just don't put this file on the USB drive.

(Do keep in mind that not all country codes are valid at this point; [see here](https://github.com/Ultimaker/opinicus/blob/d47f1813c67e589ef1cc4717cc4bfe56940e5f84/griffin/system/eeprom.py#L40-L50))
## How has this been tested
By placing the text files mentioned above on a ext4 formatted usb stick and rebooting.


## Ready for Review Checklist
To help with deciding if this PR is RFR, use this checklist.

The author confirms that:
- [x] the author has **self-reviewed** this work and is highly confident about the quality
- [x] this work satisfies all **acceptance criteria** that are stated in the linked ticket
- [ ] this work has been **tested** on all product families and the process and results are documented in the above section
- [x] The **description** above is concise yet complete
- [ ] the reviewer has been offered a **walkthrough** (if needed)
- [ ] the code is **cleaned** from any rubbish (e.g. meaningless comments, log-spamming, etc...)
- [x] remaining `#TODO` **comments** mention a Jira ticket number
- [x] all **CI** checks are passing
- [x] all **commits** are (re)structured to be meaningful and clearly arranged, and all are prepended with the ticket number for traceability

onnodb committed:
SEM-146 Write country lock during EEPROM provisioning

Extend the printer provisioning logic in the `init.sh` boot script, so
that a "locked country code" can be set in the mainboard EEPROM.

The existing provisioning code looks for an `article_number` file on a
USB drive that's inserted during boot (ext4, single partition). The
contents of that file, a single line of text in hex notation describing
the BOM number, is then written to the mainboard EEPROM using
`i2ctransfer`.

We duplicate this code, so that an additional file `country_code_lock`
can be used to also write a "locked country code" to the EEPROM, at
address 0x0118. This file is optional, so default provisioning keeps
whatever data is here, intact. The file works the same way: it should
contain a single line of text, with a hex representation of the locked
country code.

Co-authored-by: Onno Broekmans <o.broekmans@ultimaker.com>
Signed-off-by: Robin den Hertog <r.denhertog@ultimaker.com>

onnodb committed:
SEM-146 Document provisioning logic in the init.sh boot script

Add a comment to the `init.sh` boot script, explaining the logic behind
the EEPROM provisioning.

onnodb committed:
SEM-146 Boyscout: improve init.sh script

Improve the code in the `check_and_set_eeprom_data()` part of the
`init.sh` boot script. Based on feedback from @alexborro:

- No longer let the while loop span the entire code within the function.
If you analyze the purpose and flow of the while loop, it turns out it's
only ever looping to wait for the USB drive to show up under the `/dev/`
directory. All other code paths lead to either a `return` or the end of
the while loop. That makes things unnecessarily confusing.

- If an `i2ctransfer` call fails, don't exit the function completely,
but give the *other* provisioning file another chance. Do make sure that
all failures and skips are properly logged.

Co-authored-by: Alex Borro <alexborro@gmail.com>
Co-authored-by: Onno Broekmans <o.broekmans@ultimaker.com>
Co-authored-by: Robin den Hertog <r.denhertog@ultimaker.com>

alexborro commented on initramfs/init.sh:
```diff
@@ -180,42 +180,81 @@ isBootingRestoreImage()
-            continue
-        fi
+    # Get the article number from EEPROM
+    art_num=$(i2ctransfer -y 3 w2@0x57 0x01 0x00 r4)
```
I think you forgot to change the I2C Bus from 3 (S-Line) to 1 (Factor-4):
```suggestion
    art_num=$(i2ctransfer -y 1 w2@0x57 0x01 0x00 r4)
```

alexborro commented on initramfs/init.sh:
```diff
@@ -180,42 +180,81 @@ isBootingRestoreImage()
-            return 0
-        fi
+    # Get the country code lock from EEPROM
+    country_code_lock=$(i2ctransfer -y 3 w2@0x57 0x01 0x18 r2)
```
```suggestion
    country_code_lock=$(i2ctransfer -y 1 w2@0x57 0x01 0x18 r2)
```

alexborro commented on initramfs/init.sh:
```diff
@@ -180,42 +180,81 @@ isBootingRestoreImage()
         # shellcheck disable=SC2086
-        if ! i2ctransfer -y 1 w6@0x57 0x01 0x00 ${article_number}; then
-            umount "${dev}"
+        if ! i2ctransfer -y 3 w6@0x57 0x01 0x00 ${article_number}; then
```
```suggestion
        if ! i2ctransfer -y 1 w6@0x57 0x01 0x00 ${article_number}; then
```

alexborro commented on initramfs/init.sh:
```diff
@@ -180,42 +180,81 @@ isBootingRestoreImage()
+        country_code="$(cat "${country_code_lock_file}")"
+        echo "Trying to write country code lock: '${country_code}'."
+        # shellcheck disable=SC2086
+        if ! i2ctransfer -y 3 w4@0x57 0x01 0x18 ${country_code}; then
```
```suggestion
        if ! i2ctransfer -y 1 w4@0x57 0x01 0x18 ${country_code}; then
```

alexborro commented on initramfs/init.sh:
```diff
@@ -180,42 +180,81 @@ isBootingRestoreImage()
+    done
 
+    echo "Attempting to mount '${dev}'."
+    if ! mount -t f2fs,ext4,vfat,auto -o exec,noatime "${dev}" "${PROVISIONING_USB_MOUNT}"; then
```
I know the original code didnt have it, but I think it is better to mount the USB Stick as Read Only, specially because it can be FAT.
```suggestion
    if ! mount -t f2fs,ext4,vfat,auto -o ro,exec,noatime "${dev}" "${PROVISIONING_USB_MOUNT}"; then
```

robinmdh committed:
CES-2679/SEM-146: user the correct BUS fomr imx8

Use the correct i2c bus for the eeprom on imx8 and also make the mount read-only, because why not be safer.

Co-authored-by: alexborro <alexborro@gmail.com>
Co-authored-by: Robin den Hertog <robinmdh@users.noreply.github.com>

robinmdh commented:
merging without aproval because @alexborro approved and the only change was the addition of a comment line.
